cpp/modmesh/buffer/pymod/SimpleArrayCaster.hpp

```diff
@@ -0,0 +1,96 @@
+#pragma once
+/*
+ * Copyright (c) 2019, Yung-Yu Chen <yyc@solvcon.net>
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * - Redistributions of source code must retain the above copyright notice,
+ *   this list of conditions and the following disclaimer.
+ * - Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ * - Neither the name of the copyright holder nor the names of its contributors
+ *   may be used to endorse or promote products derived from this software
+ *   without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <modmesh/buffer/buffer.hpp>
+#include <pybind11/pybind11.h>
```

Comment:
> Python.h always need to be included in the first place

copied.

>  to name this file as something like common.hpp

only the files including pybind11 functions that use `SimpleArray` as parameters and return values need to include this header file. 

In addition, since no one is using the caster in this file (actually only test code uses it now), I think we can consider renaming it in the future.

>  make other files under pymod to include this one in the first place.

not sure if you are saying putting this header file into `buffer_pymod.hpp`

if yes, since `wrap_SimpleArray.hpp`  already includes  `buffer_pymod.hpp`, and when `wrap_SimpleArray.hpp` tries to include `SimpleArrayCaster.hpp`, there will be some casting issues on `SimpleArray`.

The easiest way to avoid the issues is not putting `SimpleArrayCaster.hpp` in  `buffer_pymod.hpp`. If you still think we should still do it, maybe we can leave an issue and resolve this later.

just for reference, error log if `wrap_SimpleArray.hpp` tries to include `SimpleArrayCaster.hpp` :
```
/home/mujin/modmesh/cpp/modmesh/python/common.hpp:273:29:   required from ‘static modmesh::python::WrapBase<Wrapper, Wrapped, Holder, WrappedBase>::wrapper_type& modmesh::python::WrapBase<Wrapper, Wrapped, Holder, WrappedBase>::commit(pybind11::module&, const char*, const char*) [with Wrapper = modmesh::python::WrapSimpleArray<float>; Wrapped = modmesh::SimpleArray<float>; Holder = std::unique_ptr<modmesh::SimpleArray<float>, std::default_delete<modmesh::SimpleArray<float> > >; WrappedBase = modmesh::SimpleArray<float>; modmesh::python::WrapBase<Wrapper, Wrapped, Holder, WrappedBase>::wrapper_type = modmesh::python::WrapSimpleArray<float>; pybind11::module = pybind11::module_]’
/home/mujin/modmesh/cpp/modmesh/buffer/pymod/wrap_SimpleArray.cpp:430:29:   required from here
/usr/include/pybind11/pybind11.h:249:33: error: cannot bind rvalue reference of type ‘modmesh::SimpleArrayFloat32&&’ {aka ‘modmesh::SimpleArray<float>&&’} to lvalue of type ‘modmesh::SimpleArray<float>’
  249 |                 = cast_out::cast(std::move(args_converter).template call<Return, Guard>(cap->f),
      |                   ~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  250 |                                  policy,
      |                                  ~~~~~~~
  251 |                                  call.parent);
      |                                  ~~~~~~~~~~~~
```

Reply:
We can finish this PR with the correct order of `Python.h` inclusion before making a decision for where to put this file.  We may probably get a clearer picture for how the files should be organized.

File organization is one action of refactoring and may be done separately.